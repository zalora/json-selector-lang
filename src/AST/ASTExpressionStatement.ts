// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTExpression from './ASTExpression';
import ASTStatement from './ASTStatement';
import { Token } from '../Lexer';

/**
 * ASTExpressionStatement is the base node for expression statements.
 * The "Select Expression" (e.g. `.<identifier>`) is an expression statement.
 */
class ASTExpressionStatement implements ASTStatement {
  token: Token;
  expression: ASTExpression;

  constructor(token: Token, expression: ASTExpression) {
    this.token = token;
    this.expression = expression;
  }

  getDescription(): string {
    const { description = '' } = this.expression || {};

    return description;
  }

  tokenLiteral(): string {
    const { literal = '' } = this.token || {};

    return literal;
  }
}

export default ASTExpressionStatement;
