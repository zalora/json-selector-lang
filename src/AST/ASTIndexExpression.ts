// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTExpression from './ASTExpression';
import { Token } from '../Lexer';

class ASTIndexExpression implements ASTExpression {
  token: Token;
  left: ASTExpression;
  index: ASTExpression;

  constructor(token: Token, left: ASTExpression, index: ASTExpression) {
    this.token = token;
    this.left = left;
    this.index = index;
  }

  getDescription(): string {
    const { description: leftDesc = '' } = this.left || {};
    const { description: indexDesc = '' } = this.index || {};

    if (!leftDesc || !indexDesc) {
      return '';
    }

    return `(${this.left.description}[${this.index.description}])`;
  }

  tokenLiteral(): string {
    const { literal = '' } = this.token || {};

    return literal;
  }
}

export default ASTIndexExpression;
