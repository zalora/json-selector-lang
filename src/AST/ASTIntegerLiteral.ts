// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTExpression from './ASTExpression';
import { Token } from '../Lexer';

class ASTIntegerLiteral implements ASTExpression {
  token: Token;
  value: string;

  constructor(token: Token, value: string) {
    this.token = token;
    this.value = value;
  }

  getDescription(): string {
    return this.value || '';
  }

  tokenLiteral(): string {
    const { literal = '' } = this.token || {};

    return literal;
  }
}

export default ASTIntegerLiteral;
