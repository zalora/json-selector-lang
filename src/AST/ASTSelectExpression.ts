// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTExpression from './ASTExpression';
import { Token } from '../Lexer';

class ASTSelectExpression implements ASTExpression {
  token: Token;
  key: string;

  constructor(token: Token, key: string) {
    this.token = token;
    this.key = key;
  }

  getDescription(): string {
    return `${this.token.literal} ${this.key}`;
  }

  tokenLiteral(): string {
    const { literal = '' } = this.token || {};

    return literal;
  }
}

export default ASTSelectExpression;
