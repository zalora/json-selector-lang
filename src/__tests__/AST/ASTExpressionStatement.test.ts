// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTExpressionStatement from '../../AST/ASTExpressionStatement';
import ASTIndexExpression from '../../AST/ASTIndexExpression';
import ASTIntegerLiteral from '../../AST/ASTIntegerLiteral';
import ASTSelectExpression from '../../AST/ASTSelectExpression';
import { dot, int, lbracket } from '../../tokens';

describe('ASTExpressionStatement', () => {
  const token = { type: dot, literal: '.' };
  const selectExpr = new ASTSelectExpression(token, 'data');
  const astExpSelect = new ASTExpressionStatement(token, selectExpr);

  it('tests with token input as select expression', () => {
    expect(astExpSelect).toMatchSnapshot();
  });

  it('tests if tokenLiteral returns literal of the token as select expression', () => {
    expect(astExpSelect.tokenLiteral()).toMatchSnapshot();
  });

  it('tests if toString returns returns a stringified select expression', () => {
    expect(astExpSelect.toString()).toMatchSnapshot();
  });

  const tokenDot = { type: dot, literal: '.' };
  const tokenLbracket = { type: lbracket, literal: '[' };
  const tokenInt = { type: int, literal: '2' };
  const lExpr = new ASTSelectExpression(tokenDot, 'data');
  const rExpr = new ASTIntegerLiteral(tokenInt, 2);
  const indexExpr = new ASTIndexExpression(tokenLbracket, lExpr, rExpr);
  const astExpIndex = new ASTExpressionStatement(tokenLbracket, indexExpr);

  it('tests with token input as index expression', () => {
    expect(astExpIndex).toMatchSnapshot();
  });

  it('tests if tokenLiteral returns literal of the token as index expression', () => {
    expect(astExpIndex.tokenLiteral()).toMatchSnapshot();
  });

  it('tests if toString returns returns a stringified index expression', () => {
    expect(astExpIndex.toString()).toMatchSnapshot();
  });
});
