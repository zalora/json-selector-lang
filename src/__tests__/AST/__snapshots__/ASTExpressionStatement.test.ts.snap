// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ASTExpressionStatement tests if toString returns returns a stringified index expression 1`] = `
"{
  \\"token\\": {
    \\"literal\\": \\"[\\"
  },
  \\"left\\": {
    \\"token\\": {
      \\"literal\\": \\".\\"
    },
    \\"key\\": \\"data\\"
  },
  \\"index\\": {
    \\"token\\": {
      \\"literal\\": \\"2\\"
    },
    \\"value\\": 2
  }
}"
`;

exports[`ASTExpressionStatement tests if toString returns returns a stringified select expression 1`] = `
"{
  \\"token\\": {
    \\"literal\\": \\".\\"
  },
  \\"key\\": \\"data\\"
}"
`;

exports[`ASTExpressionStatement tests if tokenLiteral returns literal of the token as index expression 1`] = `"["`;

exports[`ASTExpressionStatement tests if tokenLiteral returns literal of the token as select expression 1`] = `"."`;

exports[`ASTExpressionStatement tests with token input as index expression 1`] = `
ASTExpressionStatement {
  "expression": ASTIndexExpression {
    "index": ASTIntegerLiteral {
      "token": Object {
        "literal": "2",
        "type": Symbol(int),
      },
      "value": 2,
    },
    "left": ASTSelectExpression {
      "key": "data",
      "token": Object {
        "literal": ".",
        "type": Symbol(.),
      },
    },
    "token": Object {
      "literal": "[",
      "type": Symbol([),
    },
  },
  "token": Object {
    "literal": "[",
    "type": Symbol([),
  },
}
`;

exports[`ASTExpressionStatement tests with token input as select expression 1`] = `
ASTExpressionStatement {
  "expression": ASTSelectExpression {
    "key": "data",
    "token": Object {
      "literal": ".",
      "type": Symbol(.),
    },
  },
  "token": Object {
    "literal": ".",
    "type": Symbol(.),
  },
}
`;
